### PR TITLE
Adding the Imagenette dataset

### PR DESCRIPTION
**Pull request by BradLarson**
This adds a wrapper to import and use Jeremy Howard's [Imagenette](https://github.com/fastai/imagenette) dataset in testing image classification models within this project. It follows the template of the other image classification datasets, and can be swapped into the current benchmarking infrastructure.

This also adds an example application where MobileNet v1 is trained using Imagenette.

No on-demand data processing pipeline has been implemented in this project, so all images are loaded and kept in memory. These images are also stored as float tensors, which can lead to exhaustion of memory on systems with less than 16 GB of RAM available.

This implementation is based on work started by [Ayush Agrawal](https://github.com/Ayush517) and it addresses issue #237.

### CODE REVIEW

File: Support/Image.swift [hunk 1]
```diff
@@ -28,6 +28,13 @@ public struct Image {
 
     let imageData: ImageTensor
 
+    public var tensor: Tensor<Float> {
+        switch self.imageData {
+        case let .float(data): return data
+        case let .uint8(data): return Tensor<Float>(data)
```
Comment (brettkoonce) [inline Support/Image.swift]: UInt8, no?

Comment (BradLarson) [reply Support/Image.swift]: In this case, it is applying a selective conversion such that a Tensor containing Floats is always obtained, even from a UInt8 source. This simplifies integration with the existing image classification dataset protocols, but isn't great for memory usage. I'm experimenting with an optimization to reduce this.

Comment (brettkoonce) [reply Support/Image.swift]: 👍

File: Datasets/Imagenette/Imagenette.swift [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    if !archiveExists {
+        printError("Archive missing, downloading...")
+        do {
+            let downloadedFile = try Data(
+                contentsOf: URL(
+                    string:
+                        "https://s3.amazonaws.com/fast-ai-imageclas/imagenette\(size.suffix).tgz")!)
+            try downloadedFile.write(to: URL(fileURLWithPath: archivePath))
+        } catch {
+            printError("Could not download CIFAR dataset, error: \(error)")
```
Comment (brettkoonce) [inline Datasets/Imagenette/Imagenette.swift]: CIFAR  --> Imagenette

Comment (BradLarson) [reply Datasets/Imagenette/Imagenette.swift]: Thanks, fixed.

**Comment by texasmichelle** [comment]
I'm getting this error:

```
$ swift run -c release SqueezeNet-Imagenette
Fetching https://github.com/kylef/Commander.git
Fetching https://github.com/kylef/Spectre.git
Cloning https://github.com/kylef/Spectre.git
Resolving https://github.com/kylef/Spectre.git at 0.9.0
Cloning https://github.com/kylef/Commander.git
Resolving https://github.com/kylef/Commander.git at 0.9.1
[5/5] Linking SqueezeNet-Imagenette
Downloading Imagenette dataset...
Archive missing, downloading...
Archive downloaded, processing...
Unarchiving completed
Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value: file /swift-base/tensorflow-swift-apis/Sources/TensorFlow/Core/TensorHandle.swift, line 128
Current stack trace:
0    libswiftCore.so                    0x00007f0edaea2cb0 swift_reportError + 50
1    libswiftCore.so                    0x00007f0edaf145f0 _swift_stdlib_reportFatalErrorInFile + 115
2    libswiftCore.so                    0x00007f0edac14625 <unavailable> + 1476133
3    libswiftCore.so                    0x00007f0edac14147 <unavailable> + 1474887
4    libswiftCore.so                    0x00007f0edac13e13 <unavailable> + 1474067
5    libswiftCore.so                    0x00007f0edac137a0 _assertionFailure(_:_:file:line:flags:) + 519
6    SqueezeNet-Imagenette              0x000055c3f7475af9 <unavailable> + 203513
7    SqueezeNet-Imagenette              0x000055c3f7475bdf <unavailable> + 203743
8    SqueezeNet-Imagenette              0x000055c3f7475590 <unavailable> + 202128
9    SqueezeNet-Imagenette              0x000055c3f7470fbe <unavailable> + 184254
10   SqueezeNet-Imagenette              0x000055c3f7471b6b <unavailable> + 187243
11   SqueezeNet-Imagenette              0x000055c3f746f9d5 <unavailable> + 178645
12   SqueezeNet-Imagenette              0x000055c3f7729fb5 <unavailable> + 3039157
13   libc.so.6                          0x00007f0ecfba5ab0 __libc_start_main + 231
14   SqueezeNet-Imagenette              0x000055c3f7469b7a <unavailable> + 154490
Illegal instruction (core dumped)
$
```

File: Datasets/Imagenette/Imagenette.swift [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+            labels.append(currentLabel)
+
+            imageCount += 1
+        }
+        currentLabel += 1
+    }
+
+    let labelTensor = Tensor<Int32>(shape: [imageCount], scalars: labels)
+    let imageTensor = Tensor<Float>(
```
Comment (texasmichelle) [inline Datasets/Imagenette/Imagenette.swift]: Something in this statement isn't working.

Comment (BradLarson) [reply Datasets/Imagenette/Imagenette.swift]: Is this on a MacBook? Which toolchain version are you using?

I strongly suspect this might be due to memory exhaustion. The float version of the dataset consumes ~8 GB of RAM by itself, and I think it constructs another ~8 GB of temporary structures when building this. It can barely run on my MacBook, but it might exhaust the memory on anything less powerful than that. I'm looking at ways this could be reduced.

Comment (brettkoonce) [reply Datasets/Imagenette/Imagenette.swift]: ~20gb on my linux box all together here, fwiw

Comment (texasmichelle) [reply Datasets/Imagenette/Imagenette.swift]: Wow. I'm running on an n1-standard-4, which only has 15GB RAM. I'll try on a bigger box.

Comment (texasmichelle) [reply Datasets/Imagenette/Imagenette.swift]: You were right - it's working on an n1-highmem-8, consuming just under 16GB of RAM. It would be worth noting this in the README.

File: Datasets/Imagenette/Imagenette.swift [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            )
+        }
+    }
+
+    let downloadPath = directory.appendingPathComponent("imagenette\(size.suffix)").path
+    let directoryExists = FileManager.default.fileExists(atPath: downloadPath)
+
+    guard !directoryExists else { return }
+
+    printError("Downloading Imagenette dataset...")
```
Comment (dan-zheng) [inline Datasets/Imagenette/Imagenette.swift]: Is use of `printError` rather than `print` for non-error messages intentional?
If so, it may be clearer to rename `printError` to `printToStderr`.

Comment (BradLarson) [reply Datasets/Imagenette/Imagenette.swift]: `printError()` was introduced as a convention for printing to standard error in PR #228 , so I used the function introduced there in the same manner as has been used for other datasets. We could rethink the naming of that function and apply across all datasets in a follow-on PR, so I'll check with Denys about this.

**Comment by rickwierenga** [comment]
My Colab session crashes when I try to load the dataset (and nothing else).

Output:

```
Downloading Imagenette dataset...
Archive missing, downloading...
Archive downloaded, processing...
Unarchiving completed
tcmalloc: large alloc 1241513984 bytes == 0x68d2c000 @  0x7f397703e1e7 0x7f39758f33d2 0x7f39758f3442 0x7f3975647ab5 0x7f397567cf9d 0x7f397564b80e 0x7f397742b7ed 0x7f3977428f76 0x7f3977428a5b 0x7f39774288d5 0x7f3977396d10 0x400480 0x400480 0x7f3977497c98
tcmalloc: large alloc 2483027968 bytes == 0xb2d2c000 @  0x7f397703e1e7 0x7f39758f33d2 0x7f39758f3442 0x7f3975647ab5 0x7f397567cf9d 0x7f397564b80e 0x7f397742b7ed 0x7f3977428f76 0x7f3977428a5b 0x7f39774288d5 0x7f3977396d10 0x400480 0x400480 0x7f3977497c98
tcmalloc: large alloc 4966055936 bytes == 0x7f37857fa000 @  0x7f397703e1e7 0x7f39758f33d2 0x7f39758f3442 0x7f3975647ab5 0x7f397567cf9d 0x7f397564b80e 0x7f397742b7ed 0x7f3977428f76 0x7f3977428a5b 0x7f39774288d5 0x7f3977396d10 0x400480 0x400480 0x7f3977497c98
tcmalloc: large alloc 9932111872 bytes == 0x7f3534ffa000 @  0x7f397703e1e7 0x7f39758f33d2 0x7f39758f3442 0x7f3975647ab5 0x7f397567cf9d 0x7f397564b80e 0x7f397742b7ed 0x7f3977428f76 0x7f3977428a5b 0x7f39774288d5 0x7f3977396d10 0x400480 0x400480 0x7f3977497c98
tcmalloc: large alloc 7763632128 bytes == 0x7f3365c00000 @  0x7f3977020b6b 0x7f3977040379 0x7f393dca9bf7 0x7f393dc56c3e 0x7f393d3518d5 0x7f393d351c89 0x7f3972b21be2 0x7f3972b16e22 0x7f3972b01883 0x7f3972b23e70 0x7f3972b23e93 0x7f3972c734de 0x7f397564c5e6 0x7f3972b233bd 0x7f397742ba40 0x7f3977428f76 0x7f3977428a5b 0x7f39774288d5 0x7f3977396d10 0x400480 0x400480 0x7f3977497c98
Process killed
```

https://colab.research.google.com/drive/1WSGbPepc1q-HEfYfCjAY_4qGMxiQXjB2

**Comment by rickwierenga** [comment]
It does work with a scaled version:

```swift
let imagenette = Imagenette(inputSize: .resized160, outputSize: 160)
```

But **not** the full version:
```swift
let imagenette = Imagenette()
```

**Comment by BradLarson** [comment]
@rickwierenga - Yeah, that's due to memory exhaustion, which we see on all but the largest VMs. We're working on better pipelining for the dataset, and wanted to have this in there first as a "before" case to make sure things worked and to compare with our later optimizations. Might not be able to use this iteration of the dataset on Colab or machines with less than 16 GB of RAM.

**Comment by rickwierenga** [comment]
I was trying to train a model on Imagenette but the validation accuracy is only 0.005 and it doesn't change at all during training. I told Ayush on the Google Code In platform, but also wanted to share it here. I copied the code from my [CIFAR10 ConvNet Notebook](https://github.com/rickwierenga/s4tf-notebooks/blob/master/notebooks/CIFAR10_ConvNet.ipynb) which did work, and only changed the dataset^1. The training accuracy is very close to 10% and it also doesn't change at all during training. Are we sure all labels are loaded correctly? 10% is exactly the accuracy you would get by guessing randomly.

Here is the Colab: https://colab.research.google.com/drive/1WhcGye-Tqjzp0NMQgUUfK6e7IMNc5B8l

^1 I'm planning on improving the model for this dataset, but wanted to get something working first.

**Comment by BradLarson** [comment]
@rickwierenga - It's possible that there's something off in the labels or other part of the dataset, but the dataset trains with MobileNet and with other networks in my tests. The model not converging could be due to problems with the input shape, how the model is initialized, the learning rate, or a number of factors.

If you swapped in this dataset for CIFAR-10, are you sure that the images are being downsampled properly to be loaded by your network that originally was taking in 32x32 images? Was the original network designed around a 0.0 - 1.0 colorspace range, or 0.0 - 255.0? Was normalization used in your original implementation? All of the above can play a role in a network not training in the same way. It can be a tricky thing to debug.
